File: Python/DeQueue_MonotonicQueue_Hard_SlidingWindowMaximum_239.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        dq = deque()
        answer = []

        for i in range(0, len(nums)):
            if dq and dq[0] <= i - k:
                dq.popleft()
            while dq and nums[i] >= nums[dq[-1]]:
                dq.pop()

            dq.append(i)

            if i >= k - 1:
                answer.append(nums[dq[0]])

        return answer    

File: C++/DeQueue_MonotonicQueue_Hard_SlidingWindowMaximum_239.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) 
    {
        deque<int> dq;
        vector<int> answer;

        for(int i = 0; i < nums.size(); i++)
        {
            if(!dq.empty() && dq.front() <= i - k)
            {
                dq.pop_front();
            }
            while(!dq.empty() and nums[i] >= nums[dq.back()])
            {
                dq.pop_back();
            }

            dq.push_back(i);
            
            if(i >= k - 1)
            {
                answer.push_back(nums[dq.front()]);
            }
        }    

        return answer;
    }
};